chainer/functions/math/sparse_matmul.py

```diff
@@ -0,0 +1,429 @@
+        if (i_b >= nb) {
+            continue;
+        }
+        int i_k = A_col[i_A];
+        if (i_k < 0) {
+            continue;
+        }
+        assert(i_k < _k);
+        int i_m = A_row[i_A];
+        if (i_m < 0) {
+            continue;
+        }
+        assert(i_m < _m);
+        int i_B = i_n + _n * (i_k + _k * i_b);
+        int i_C = i_n + _n * (i_m + _m * i_b);
+        A val_A = A_data[i_A];
+        C val = _B[i_B] * val_A;
+        atomicAdd(&_C[i_C], val);
+        ''',
+        'sparse_matmul')
+
+
+class SparseMatMul(function_node.FunctionNode):
+
+    def __init__(self, sp_row, sp_col, sp_shape,
+                 transa=False, transb=False, transc=False,
+                 dtype=None):
+        if sp_row.ndim != sp_col.ndim:
+            raise ValueError('ndim of sp_row and sp_col must be the same.')
+        if sp_row.ndim != 1 and sp_row.ndim != 2:
+            raise ValueError('ndim of sp_row and sp_col must be one or two.')
+        for i in range(sp_row.ndim):
+            if sp_row.shape[i] != sp_col.shape[i]:
+                _msg = 'shape of sp_row and sp_col must be the same.'
+                raise ValueError(_msg)
+        if len(sp_shape) != 2:
+            raise ValueError('len(sp_shape) must be two.')
+        self.sp_row = sp_row  # ((nb,) ldnz)
+        self.sp_col = sp_col  # ((nb,) ldnz)
+        self.sp_shape = sp_shape  # (_m, _k) when transa is False
+        self.transa = transa
+        self.transb = transb
+        self.transc = transc
+        self.dtype = dtype
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        sp_type, dn_type = in_types
+        # sp_type.shape: ((nb,) ldnz)
+        # dn_type.shape: ((nb,) _k, _n) when transb is False
+        sp_k_axis = -1
+        if self.transa:
+            sp_k_axis = -2
+        dn_k_axis = -2
+        if self.transb:
+            dn_k_axis = -1
+        type_check.expect(
+            sp_type.dtype.kind == 'f',
+            dn_type.dtype.kind == 'f',
+            dn_type.ndim >= 2,
+            dn_type.ndim <= 3,
+            sp_type.ndim == dn_type.ndim - 1,
+            sp_type.shape[-1] == self.sp_row.shape[-1],
+            self.sp_shape[sp_k_axis] == dn_type.shape[dn_k_axis],
+        )
+        dn_ndim = type_check.eval(dn_type.ndim)
+        if dn_ndim == 3:
+            type_check.expect(
+                sp_type.shape[0] == self.sp_row.shape[0],
+                dn_type.shape[0] == self.sp_row.shape[0],
+            )
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1))
+        sp, dn = inputs
+        c = _sparse_matmul(sp, self.sp_row, self.sp_col, self.sp_shape, dn,
+                           self.transa, self.transb, self.transc, self.dtype)
+        return utils.force_array(c, self.dtype),
+
+    def backward(self, indexes, grad_outputs):
+        sp, dn = self.get_retained_inputs()
+        g_c, = grad_outputs
+
+        g_sp = None
+        if 0 in indexes:
```

Comment:
We usually use follow this idiom when checking the indices. Could you consider using it? https://github.com/chainer/chainer/blob/c2ad311fb043d7fb37acdd97f473651637df595c/chainer/functions/math/trigonometric.py#L49-L54

Reply:
OK, I will edit code to conform that style.